server/src/main/java/org/opensearch/action/search/SearchRequestSlowLog.java

```diff
@@ -0,0 +1,265 @@
+        TimeValue.timeValueNanos(-1),
+        TimeValue.timeValueMillis(-1),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    public static final Setting<TimeValue> CLUSTER_SEARCH_REQUEST_SLOWLOG_THRESHOLD_DEBUG_SETTING = Setting.timeSetting(
+        CLUSTER_SEARCH_REQUEST_SLOWLOG_PREFIX + ".threshold.debug",
+        TimeValue.timeValueNanos(-1),
+        TimeValue.timeValueMillis(-1),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    public static final Setting<TimeValue> CLUSTER_SEARCH_REQUEST_SLOWLOG_THRESHOLD_TRACE_SETTING = Setting.timeSetting(
+        CLUSTER_SEARCH_REQUEST_SLOWLOG_PREFIX + ".threshold.trace",
+        TimeValue.timeValueNanos(-1),
+        TimeValue.timeValueMillis(-1),
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+    public static final Setting<SlowLogLevel> CLUSTER_SEARCH_REQUEST_SLOWLOG_LEVEL = new Setting<>(
+        CLUSTER_SEARCH_REQUEST_SLOWLOG_PREFIX + ".level",
+        SlowLogLevel.TRACE.name(),
+        SlowLogLevel::parse,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    private static final ToXContent.Params FORMAT_PARAMS = new ToXContent.MapParams(Collections.singletonMap("pretty", "false"));
+    private SlowLogLevel level;
+
+    public SearchRequestSlowLog(ClusterService clusterService) {
+        this.logger = LogManager.getLogger(CLUSTER_SEARCH_REQUEST_SLOWLOG_PREFIX);
+        Loggers.setLevel(this.logger, SlowLogLevel.TRACE.name());
+
+        this.warnThreshold = clusterService.getClusterSettings().get(CLUSTER_SEARCH_REQUEST_SLOWLOG_THRESHOLD_WARN_SETTING).nanos();
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(CLUSTER_SEARCH_REQUEST_SLOWLOG_THRESHOLD_WARN_SETTING, this::setWarnThreshold);
+        this.infoThreshold = clusterService.getClusterSettings().get(CLUSTER_SEARCH_REQUEST_SLOWLOG_THRESHOLD_INFO_SETTING).nanos();
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(CLUSTER_SEARCH_REQUEST_SLOWLOG_THRESHOLD_INFO_SETTING, this::setInfoThreshold);
+        this.debugThreshold = clusterService.getClusterSettings().get(CLUSTER_SEARCH_REQUEST_SLOWLOG_THRESHOLD_DEBUG_SETTING).nanos();
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(CLUSTER_SEARCH_REQUEST_SLOWLOG_THRESHOLD_DEBUG_SETTING, this::setDebugThreshold);
+        this.traceThreshold = clusterService.getClusterSettings().get(CLUSTER_SEARCH_REQUEST_SLOWLOG_THRESHOLD_TRACE_SETTING).nanos();
+        clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(CLUSTER_SEARCH_REQUEST_SLOWLOG_THRESHOLD_TRACE_SETTING, this::setTraceThreshold);
+
+        this.level = clusterService.getClusterSettings().get(CLUSTER_SEARCH_REQUEST_SLOWLOG_LEVEL);
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(CLUSTER_SEARCH_REQUEST_SLOWLOG_LEVEL, this::setLevel);
```

Comment:
Adding these settings update consumer for each search request level object creation can cause:

- SearchRequestSlowLog object memory leaks as uncleared references to this:: functions from settings update consumer
- The time taken for processing settings update will grow significantly with time as all leaked consumers will be invoked

We can resolve this issue by creating single node level object or as you suggested offline, not adding these consumers upon each search request

Reply:
@jainankitk good catch. I confirmed that updateConsumer causes `SearchRequestSlowLog` objects to pile up. 

After 1000 queries, 1001 SearchRequestSlowLog objects exist in the heap:
```
% jmap -histo:live 91904 | grep SearchRequestSlowLog
  57:          1001          64512  org.opensearch.action.search.SearchRequestSlowLog
 123:          1001          16128  org.opensearch.action.search.SearchRequestSlowLog$$Lambda$3865/0x0000000801ce7840
 124:          1001          16128  org.opensearch.action.search.SearchRequestSlowLog$$Lambda$3866/0x0000000801ce6440
 125:          1001          16128  org.opensearch.action.search.SearchRequestSlowLog$$Lambda$3867/0x0000000801ce5040
 126:          1001          16128  org.opensearch.action.search.SearchRequestSlowLog$$Lambda$3868/0x0000000801ce4040
 127:          1008          16128  org.opensearch.action.search.SearchRequestSlowLog$$Lambda$3869/0x0000000801ce2040
4769:             1             16  org.opensearch.action.search.SearchRequestSlowLog$$Lambda$971/0x0000000801594440
```

I removed addSettingsUpdateConsumer() for now since they are not needed per request. After 1000 queries just 1 object remains:
```
% jmap -histo:live 4944 | grep SearchRequestSlowLog
4254:             1             16  org.opensearch.action.search.SearchRequestSlowLog$$Lambda$971/0x000000080158c440
```

Reply:
@dzane17 - Thank you for this experiment and confirming the correctness of our understanding.